File: frontend/app/types/events.ts
/**
 * app/types/events.ts
 *
 * TypeScript interfaces for every WebSocket message type in the
 * Wafrivet Field Vet Live streaming protocol (Phase 5 contract).
 *
 * Source of truth: backend/streaming/events.py
 * WebSocket endpoint: /ws/{user_id}/{session_id}
 *
 * ── Transport directions ──────────────────────────────────────────────────────
 *  Server → Client
 *    Binary frame  : raw 24 kHz PCM audio from Gemini (play via Web Audio)
 *    JSON frame    : one of ServerEvent (discriminated on `type`)
 *
 *  Client → Server
 *    Binary frame  : raw 16 kHz 16-bit mono PCM audio from the mic
 *    JSON frame    : one of ClientMessage (IMAGE | TEXT)
 */

// ── Event type string constants ───────────────────────────────────────────────
//
// These string literals MUST match the Python constants in events.py exactly.

/** WebSocket path prefix. Full path: /ws/{user_id}/{session_id} */
export const WS_PATH = '/ws' as const;

/** All stable event type string literals. */
export const WT = {
  // ── Server → Client ────────────────────────────────────────────────────────
  /** Discard the browser audio queue; stop playing stale audio. */
  AUDIO_FLUSH:           'AUDIO_FLUSH',
  /** STT transcription fragment (user voice or agent TTS output). */
  TRANSCRIPTION:         'TRANSCRIPTION',
  /** Agent finished its response turn. */
  TURN_COMPLETE:         'TURN_COMPLETE',
  /** The ADK recommend_products or search_products tool fired: render product cards. */
  PRODUCTS_RECOMMENDED:  'PRODUCTS_RECOMMENDED',
  /** The ADK manage_cart or update_cart tool fired: refresh cart badge and line items. */
  CART_UPDATED:          'CART_UPDATED',
  /** The ADK generate_checkout_link tool fired: show Pay Now button. */
  CHECKOUT_LINK:         'CHECKOUT_LINK',
  /** The ADK update_location tool fired: state confirmed. */
  LOCATION_CONFIRMED:    'LOCATION_CONFIRMED',
  /** The ADK find_nearest_vet_clinic tool fired: render clinic cards. */
  CLINICS_FOUND:         'CLINICS_FOUND',
  /** An ADK tool returned a non-success status. */
  TOOL_ERROR:            'TOOL_ERROR',
  /** Terminal Gemini model error (safety, block, token limit, cancel). */
  ERROR:                 'ERROR',
  /** Phase 3: place_order confirmed; show reference number + SMS notice. */
  ORDER_CONFIRMED:       'ORDER_CONFIRMED',
  /** Phase 3: identify_product_from_frame is active; show scanning indicator. */
  SCANNING_PRODUCT:      'SCANNING_PRODUCT',

  // ── Client → Server (JSON frames) ─────────────────────────────────────────
  /** A JPEG camera frame, base64-encoded. */
  IMAGE:     'IMAGE',
  /** A text message (typed fallback or command). */
  TEXT:      'TEXT',
  /** Signal the backend to interrupt the current AI response turn. */
  INTERRUPT: 'INTERRUPT',
  /** Browser GPS coordinates sent once geolocation resolves. */
  LOCATION_DATA: 'LOCATION_DATA',
} as const;

export type WTKey = keyof typeof WT;

// ── Domain model shapes ───────────────────────────────────────────────────────

/**
 * A veterinary product card pushed by the recommend_products ADK tool.
 *
 * Field names match the Supabase `products` table columns selected in
 * backend/agent/tools/products.py:
 *   SELECT id, name, base_price, image_url, description, dosage_notes
 *
 * NOTE: the price field is `base_price` (the DB column name), not `price_ngn`.
 *       Prices are in Nigerian Naira (NGN).
 *
 * image_url is a relative path served from the Next.js public/ directory,
 * e.g. "/images/products/BLT-001.jpg". Place product images in
 * frontend/public/images/products/ for local development.
 */
export interface Product {
  id: string;
  name: string;
  /** Fallback price in NGN from the products table `base_price` column. */
  base_price: number;
  /**
   * Distributor-specific price in NGN returned by hybrid_search_products RPC.
   * Takes priority over base_price when present.
   */
  price?: number;
  /** Reciprocal Rank Fusion score; lower = better match. Results already arrive sorted. */
  rrf_rank?: number;
  /** Distributor UUID that supplied this price; null for products without a local distributor. */
  distributor_id?: string | null;
  /** Relative path e.g. "/images/products/BLT-001.jpg" (Next.js public/). */
  image_url: string;
  description: string;
  dosage_notes: string;
}

/**
 * A veterinary clinic returned by the find_nearest_vet_clinic ADK tool.
 * Matches the _normalise_clinic() shape in backend/agent/tools/vet_clinics.py.
 */
export interface Clinic {
  name: string;
  address: string;
  phone: string | null;
  /** True if the clinic is currently open, null if hours data is unavailable. */
  openNow: boolean | null;
  /** Direct Google Maps link for in-app navigation. */
  googleMapsUri: string | null;
  lat: number | null;
  lon: number | null;
}

/**
 * A single line item in the farmer's active cart.
 * Matches backend/agent/session.py CartItem TypedDict exactly.
 */
export interface CartItem {
  product_id: string;
  product_name: string;
  quantity: number;
  /** Unit price in NGN. */
  unit_price: number;
  /** unit_price × quantity in NGN. */
  subtotal: number;
}

// ── Server → Client JSON events ───────────────────────────────────────────────

/** Instructs the browser to immediately discard its audio playback queue. */
export interface AudioFlushEvent {
  type: 'AUDIO_FLUSH';
}

/**
 * A speech-to-text transcription fragment.
 * author = "user" for farmer microphone input; agent name for AI output.
 * is_final = true when the fragment is complete (not a streaming partial).
 */
export interface TranscriptionEvent {
  type: 'TRANSCRIPTION';
  text: string;
  /** "user" | agent name (e.g. "Wafrivet Field Vet") */
  author: string;
  is_final: boolean;
}

/** Agent completed its response turn; hide any "thinking" indicator. */
export interface TurnCompleteEvent {
  type: 'TURN_COMPLETE';
}

/** Render product cards in the UI when the AI recommends treatments. */
export interface ProductsRecommendedEvent {
  type: 'PRODUCTS_RECOMMENDED';
  products: Product[];
  /** Optional human-readable summary from the tool. */
  message: string;
}

/** Full cart state refresh. Update badge count and line-item list. */
export interface CartUpdatedEvent {
  type: 'CART_UPDATED';
  items: CartItem[];
  /** Current cart total in NGN. */
  cart_total: number;
  message: string;
}

/**
 * Paystack payment link generated by generate_checkout_link.
 * Render a "Pay Now" button that opens checkout_url in a new tab.
 */
export interface CheckoutLinkEvent {
  type: 'CHECKOUT_LINK';
  /** Paystack authorization URL (test or live). */
  checkout_url: string;
  /** Unique Paystack transaction reference for lookup/reconciliation. */
  payment_reference: string;
  message: string;
}

/** Farmer's Nigerian state confirmed; use for product region filtering. */
export interface LocationConfirmedEvent {
  type: 'LOCATION_CONFIRMED';
  /** Canonical title-case state name e.g. "Rivers", "Lagos", "FCT". */
  state: string;
  message: string;
}

/**
 * Nearest vet clinic results from find_nearest_vet_clinic ADK tool.
 * Render as ClinicCardRow below the product strip.
 */
export interface ClinicsFoundEvent {
  type: 'CLINICS_FOUND';
  clinics: Clinic[];
  /** Effective search radius in metres used for this response. */
  radius_m: number;
  /**
   * Non-null when no clinics were found within 50 km.
   * Fatima speaks this aloud and it is displayed in the card row.
   */
  fallback_message: string | null;
  message: string;
}

/** An ADK tool returned a non-success status; agent will narrate the error. */
export interface ToolErrorEvent {
  type: 'TOOL_ERROR';
  tool_name: string;
  error: string;
}

/**
 * Phase 3: place_order confirmed.
 * Display the order reference number and an SMS confirmation notice.
 */
export interface OrderConfirmedEvent {
  type: 'ORDER_CONFIRMED';
  /** WV-XXXXXX reference string the farmer should keep. */
  order_reference: string;
  /** Total amount charged in NGN. */
  total: number;
  /** Cart line items at time of confirmation. */
  items: CartItem[];
  /** Human-readable delivery window e.g. "24–48 hours". */
  estimated_delivery: string;
  /** True if Termii SMS was successfully dispatched. */
  sms_sent: boolean;
  message: string;
}

/**
 * Phase 3: identify_product_from_frame is in progress.
 * Show a scanning indicator on the camera overlay until PRODUCTS_RECOMMENDED arrives.
 */
export interface ScanningProductEvent {
  type: 'SCANNING_PRODUCT';
  message: string;
}

/**
 * Terminal Gemini model error. Possible codes:
 *   SAFETY | PROHIBITED_CONTENT | BLOCKLIST | MAX_TOKENS | CANCELLED
 * The WebSocket will close after this event.
 */
export interface ModelErrorEvent {
  type: 'ERROR';
  code: string;
  message: string;
}

/**
 * Sent by the bridge immediately before AUDIO_FLUSH when a barge-in occurs.
 * The frontend only needs to acknowledge it; AUDIO_FLUSH handles the queue.
 */
export interface InterruptedEvent {
  type: 'interrupted';
}

/** Discriminated union of every server → client JSON event. */
export type ServerEvent =
  | AudioFlushEvent
  | TranscriptionEvent
  | TurnCompleteEvent
  | ProductsRecommendedEvent
  | CartUpdatedEvent
  | CheckoutLinkEvent
  | LocationConfirmedEvent
  | ClinicsFoundEvent
  | ToolErrorEvent
  | ModelErrorEvent
  | InterruptedEvent
  | OrderConfirmedEvent
  | ScanningProductEvent;

// ── Client → Server JSON messages ─────────────────────────────────────────────
// Binary PCM audio frames are sent as ArrayBuffer — not typed here.

/**
 * A JPEG video frame captured from the rear camera.
 * data is the raw base64 string (no data: URI prefix).
 */
export interface ImageMessage {
  type: 'IMAGE';
  /** Base64-encoded JPEG, no "data:image/jpeg;base64," prefix. */
  data: string;
}

/** A text message: typed user input or programmatic command. */
export interface TextMessage {
  type: 'TEXT';
  text: string;
}

/**
 * Sent to interrupt an in-progress AI response turn.
 * The backend bridge stops the current Gemini turn and the server
 * emits AUDIO_FLUSH to clear the browser's audio queue.
 */
export interface InterruptMessage {
  type: 'INTERRUPT';
}

/**
 * GPS coordinates captured by the browser's Geolocation API.
 * Sent once after connection to write farmer_lat / farmer_lon into the
 * ADK session state so find_nearest_vet_clinic can use them.
 */
export interface LocationDataMessage {
  type: 'LOCATION_DATA';
  lat: number;
  lon: number;
  /** LGA (Local Government Area) resolved by the geocode API route. */
  lga?: string;
  /** Canonical Nigerian state name resolved by the geocode API route. */
  state?: string;
}

/** Discriminated union of all client → server JSON messages. */
export type ClientMessage = ImageMessage | TextMessage | InterruptMessage | LocationDataMessage;

// ── Type guard ────────────────────────────────────────────────────────────────

/**
 * Minimal runtime guard: confirms the parsed JSON has a string `type` field.
 * Full schema validation is omitted for performance (trust the backend).
 */
export function isServerEvent(data: unknown): data is ServerEvent {
  return (
    typeof data === 'object' &&
    data !== null &&
    'type' in data &&
    typeof (data as Record<string, unknown>).type === 'string'
  );
}

File: backend/streaming/events.py
"""
backend/streaming/events.py

Typed outbound event envelopes sent over the WebSocket to the frontend.

Every message is a JSON object with a top-level ``"type"`` key drawn from the
constants below, plus the fields defined in each TypedDict.

Phase 5 contract — these type strings are stable:
    AUDIO_FLUSH            — browser must discard its audio queue immediately
    TRANSCRIPTION          — speech-to-text for user or agent speech
    TURN_COMPLETE          — agent has finished its response turn
    PRODUCTS_RECOMMENDED   — product cards to render in the UI
    CART_UPDATED           — shopping-cart state refresh
    CHECKOUT_LINK          — payment link to present to the farmer
    LOCATION_CONFIRMED     — Nigerian state confirmed in session
    TOOL_ERROR             — a tool call failed; no data available

Phase 3 extensions:
    ORDER_CONFIRMED        — order placed successfully; show reference + SMS notice
    SCANNING_PRODUCT       — camera product identification in progress
"""
from __future__ import annotations

from typing import Any, List, Optional


# ---------------------------------------------------------------------------
# Type constant strings (stable Phase 5 contract)
# ---------------------------------------------------------------------------

T_AUDIO_FLUSH          = "AUDIO_FLUSH"
T_TRANSCRIPTION        = "TRANSCRIPTION"
T_TURN_COMPLETE        = "TURN_COMPLETE"
T_PRODUCTS_RECOMMENDED = "PRODUCTS_RECOMMENDED"
T_CART_UPDATED         = "CART_UPDATED"
T_CHECKOUT_LINK        = "CHECKOUT_LINK"
T_LOCATION_CONFIRMED   = "LOCATION_CONFIRMED"
T_CLINICS_FOUND        = "CLINICS_FOUND"
T_TOOL_ERROR           = "TOOL_ERROR"
# Phase 3
T_ORDER_CONFIRMED      = "ORDER_CONFIRMED"
T_SCANNING_PRODUCT     = "SCANNING_PRODUCT"


# ---------------------------------------------------------------------------
# Factory helpers — each returns a dict ready for websocket.send_json()
# ---------------------------------------------------------------------------

def audio_flush_event() -> dict:
    return {"type": T_AUDIO_FLUSH}


def transcription_event(text: str, author: str, is_final: bool = False) -> dict:
    """
    Args:
        text    — the transcribed text fragment
        author  — "user" for input transcription, agent name for output
        is_final — True when this is the final (non-partial) transcription
    """
    return {
        "type":     T_TRANSCRIPTION,
        "text":     text,
        "author":   author,
        "is_final": is_final,
    }


def turn_complete_event() -> dict:
    return {"type": T_TURN_COMPLETE}


def products_recommended_event(products: List[dict], message: str = "") -> dict:
    """
    Args:
        products — list of product dicts from recommend_products tool response:
                   [{id, name, price_ngn, image_url, description, dosage_notes}]
        message  — optional human-readable summary from the tool
    """
    return {
        "type":     T_PRODUCTS_RECOMMENDED,
        "products": products,
        "message":  message,
    }


def cart_updated_event(
    items: List[dict],
    cart_total: float,
    message: str = "",
) -> dict:
    """
    Args:
        items      — list of cart line items from manage_cart tool response
        cart_total — current cart total in NGN
        message    — optional human-readable summary from the tool
    """
    return {
        "type":       T_CART_UPDATED,
        "items":      items,
        "cart_total": cart_total,
        "message":    message,
    }


def checkout_link_event(
    checkout_url: str,
    payment_reference: str,
    message: str = "",
) -> dict:
    return {
        "type":              T_CHECKOUT_LINK,
        "checkout_url":      checkout_url,
        "payment_reference": payment_reference,
        "message":           message,
    }


def location_confirmed_event(state: str, message: str = "") -> dict:
    return {
        "type":    T_LOCATION_CONFIRMED,
        "state":   state,
        "message": message,
    }


def clinics_found_event(
    clinics: List[dict],
    radius_m: int,
    fallback_message: Optional[str] = None,
    message: str = "",
) -> dict:
    """
    Args:
        clinics          — list of clinic dicts from find_nearest_vet_clinic:
                           [{name, address, phone, openNow, googleMapsUri, lat, lon}]
        radius_m         — the effective search radius used, in metres
        fallback_message — non-None when no clinics were found within 50 km;
                           Fatima speaks this and the frontend shows a fallback card
        message          — optional human-readable summary from the tool
    """
    return {
        "type":             T_CLINICS_FOUND,
        "clinics":          clinics,
        "radius_m":         radius_m,
        "fallback_message": fallback_message,
        "message":          message,
    }


def tool_error_event(tool_name: str, error: str) -> dict:
    return {
        "type":      T_TOOL_ERROR,
        "tool_name": tool_name,
        "error":     error,
    }


# ---------------------------------------------------------------------------
# Phase 3 factory helpers
# ---------------------------------------------------------------------------

def order_confirmed_event(
    order_reference: str,
    total: float,
    items: List[dict],
    estimated_delivery: str,
    sms_sent: bool = False,
    message: str = "",
) -> dict:
    """
    Args:
        order_reference    — WV-XXXXXX reference string
        total              — order total in NGN
        items              — list of cart line items at time of confirmation
        estimated_delivery — human-readable delivery window string
        sms_sent           — True if Termii SMS was dispatched successfully
        message            — optional human-readable summary
    """
    return {
        "type":               T_ORDER_CONFIRMED,
        "order_reference":    order_reference,
        "total":              total,
        "items":              items,
        "estimated_delivery": estimated_delivery,
        "sms_sent":           sms_sent,
        "message":            message,
    }


def scanning_product_event(message: str = "") -> dict:
    """
    Emitted when identify_product_from_frame is called.
    Tells the frontend to show a scanning indicator on the camera overlay.
    """
    return {
        "type":    T_SCANNING_PRODUCT,
        "message": message or "Scanning product label…",
    }
